lib/Dialect/FIRRTL/Transforms/AddSeqMemPorts.cpp

```diff
@@ -324,6 +342,8 @@ void AddSeqMemPortsPass::createOutputFile(hw::HWModuleLike module) {
 void AddSeqMemPortsPass::runOnOperation() {
   auto *context = &getContext();
   auto circuit = getOperation();
+  AnnotationSet dontTouchAnno(circuit.getContext());
+  dontTouch = &dontTouchAnno;
```

Comment:
We have a context already (no need to fetch it from the circuit).

You could move this inside `if (userPorts.size() > 0) {` so you don't bother constructing it if the pass is doing nothing, and then add the DontTouch to the AnnoSet immediately instead of lazily. Although it might be a bit of a footgun to have an uninitialized ptr sitting in the members.

One annoying thing is that creating the annotation set this way will first construct an empty dictionary, but it's not a big loss.  To make it faster you could construct it right from a DontTouch attribute:

```c++
// Hopefully enough implicit conversion to go from DictionaryAttr to ArrayRef<Attribute>
AnnotationSet dontTouchAnno(DictionaryAttr::get(
      context, {{StringAttr::get(context, "class"),
                      StringAttr::get(context, dontTouchAnnoClass)}}))

```

Can we get away with annotating only the ports on the DUT? If so then we can just keep this anno set local to this function.

Reply:
Good observation! It would be sufficient to only put the annotations on the DUT. That would avoid the dance with the `AnnotationSet` pointer.

Reply:
Switched to only applying this to the DUT (and also to the DUT instance ports due to how the pass works).